[lldb-dap] Improve `stackTrace` and `exceptionInfo` DAP request handlers

ashgti opened the pull request:
Refactoring `stackTrace` to perform frame look ups in a more on-demand fashion to improve overall performance.

Additionally adding additional information to the `exceptionInfo` request to report exception stacks there instead of merging the exception stack into the stack trace. The `exceptionInfo` request is only called if a stop event occurs with `reason='exception'`, which should mitigate the performance of `SBThread::GetCurrentException` calls.

Adding unit tests for exception handling and stack trace supporting.

ashgti committed:
[lldb-dap] Improve `stackTrace` and `exceptionInfo` DAP request handlers.

Refactoring `stackTrace` to perform frame look ups in a more on-demand fashion to improve overall performance.

Additionally adding additional information to the `exceptionInfo` request to report exception stacks there instead of merging the exception stack into the stack trace. The `exceptionInfo` request is only called if a stop event occurs with `reason='exception'`, which should mitigate the performance of `SBThread::GetCurrentException` calls.

Adding unit tests for exception handling and stack trace supporting.

ashgti commented:
This was based on the comments in https://github.com/llvm/llvm-project/pull/104874

LMKWYT

ashgti committed:
Move the ExceptionDetails type info closer to request_exceptionInfo

ashgti committed:
Cleaning up unused variables.

ashgti committed:
Cleanup debug logs

walter-erquinigo reviewed (approved):
This LGTM, and thank you for your diligence with writing tests.

ashgti commented:
@jeffreytan81 I moved the expensive call into the `exceptionInfo` request, this is only going to be triggered if a thread has the stop event = 'exception', so it shouldn't be triggered while stepping. This does remove the option you added in the prior commit, since I think this isolated the expensive call to a place where the additional information it provides justifies the call and is only called in specific circumstances.

LMKWYT

jeffreytan81 commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -3106,80 +3228,105 @@ void request_stackTrace(const llvm::json::Object &request) {
-    body.try_emplace("totalFrames", totalFrames);
+    // If we loaded all the frames, set the total frame to the current total,
+    // otherwise use the totalFrames to indiciate more data is available.
+    body.try_emplace("totalFrames",
```
Have you tested this? Since you are returning one more frame, I am not sure if this is causing VSCode client to ask for one more frame when you click "Get More frames" link. I believe @clayborg  suggested defining a page size (like 20 frames), then if it is not done, return  `startFrame + stackFrames.size() + STACK_PAGE_SIZE` instead. Then, client will ask for `STACK_PAGE_SIZE` instead of 1 more frame.

ashgti replied:
Done.

I had tested this but the 20 page size does work better with deep stack traces.

Also updated the tests to verify this as well.

ashgti committed:
Adjusting the lldb-dap stackTrace to use a StackPageSize=20 const instead of only a +1.

This helps improve pagination of deep stack traces.

ashgti committed:
Applying clang-format and python format.

jeffreytan81 commented:
@ashgti, is there any concern to guard this feature behind a launch configuration option? I am happy original profile trace hot path is moved to a rare DAP request, but I am not sure about the cost of `GetExtendedBacktraceThread()` and if it will show up as next hot path in the new profile trace. I think it would be safer to make this feature opt-in instead of enabling by default. For example, the newly reported perf issue on Windows (https://github.com/msys2/MINGW-packages/pull/21777) may or may not be related with this feature.

Also, if I understand it correctly, the extended backtrace is a plugin style feature which any runtime can implement its own extended backtrace however expensive it is. This means we can't foresee the future cost of enabling this. Adding a knob for it would be useful for people to quickly opt in/out.

ashgti committed:
Only display the extended stack trace if `enableDisplayExtendedBacktrace` is true and document `enableDisplayExtendedBacktrace`.

ashgti commented:
Reverted the removal of `enableDisplayExtendedBacktrace` and added `enableDisplayExtendedBacktrace` to the documentation and package.json description of the lldb-dap debugger.

jeffreytan81 commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -3106,80 +3235,109 @@ void request_stackTrace(const llvm::json::Object &request) {
+      // *NOTE*: Threads can be chained across mutliple backtraces, so we
+      // need to keep track of each backtrace we've traversed fully in the
+      // offset.
+      while (!frame.IsValid() && current.IsValid() && !threadCluster.empty()) {
```
Nit: it may worth moving extended back trace specific code into a dedicated helper function to improve readability.

ashgti replied:
Sorry for the delay, I was out for a week.

I refactored the `request_stackTrace` implementation and added some additional tests to validate the pagination behavior and ensure stack traces across thread backtraces are handled correctly.

ashgti committed:
Restructuring the stacktrace generation code to use more helpers and to simplify the implementation.

ashgti committed:
Apply clang-format.
